lib/src/clojure_lsp/feature/restructure_keys.clj

```diff
@@ -0,0 +1,190 @@
+  (reduce z/append-child
+          (z/of-node (n/list-node []))
+          children))
+
+(defn ^:private z-of-map [map-entries]
+  (reduce (fn [loc [k v]]
+            (z/assoc loc k v))
+          (z/of-node (n/map-node []))
+          map-entries))
+
+(defn ^:private z-of-symbol [element-name]
+  (z/of-node (n/token-node element-name)))
+
+(defn ^:private loc-kw-name [key-loc]
+  (let [key-sexpr (z/sexpr key-loc)]
+    (when (keyword? key-sexpr)
+      (name key-sexpr))))
+
+(defn ^:private reference-elems [local-node db filename]
+  (let [local-meta (meta local-node)
+        elem (q/find-local-under-cursor db filename (:row local-meta) (:col local-meta))]
+    (q/find-references db elem false)))
+
+(defn ^:private restructure-data [key-loc val-loc
+                                  db filename
+                                  {:keys [map-prefix map-auto-resolved?]}]
+  (let [key-sexpr (z/sexpr key-loc)]
+    (cond
+      ;; {:keys [a]} and all its variations
+      (and (keyword? key-sexpr)
+           (contains? #{"keys" "syms"} (name key-sexpr)))
+      (let [k-prefix (namespace (:k (z/node key-loc)))
+            ;; #:prefix{:_/keys [a]} -> (:a element)
+            abort-default-ns? (= k-prefix "_")
+            default-ns (when-not abort-default-ns?
+                         (or k-prefix map-prefix))
+            default-ns-auto-resolved? (when-not abort-default-ns?
+                                        (or (:auto-resolved? (z/node key-loc))
+                                            map-auto-resolved?))
+            replace-with (case (name key-sexpr)
+                           "keys"
+                           (fn [local-node]
+                             (let [local-sexpr (n/sexpr local-node)]
+                               (cond
+                                 ;; {:keys [::a]}            -> (::a element)
+                                 ;; {:keys [::my-alias/a]}   -> (::my-alias/a element)
+                                 (:auto-resolved? local-node)   local-node
+                                 ;; {:keys [my-ns/a]}        -> (:my-ns/a element)
+                                 ;; {:keys [:my-ns/a]}       -> (:my-ns/a element)
+                                 (qualified-ident? local-sexpr) (n/keyword-node (keyword local-sexpr))
+                                 ;; {:my-ns/keys [a]}        -> (:my-ns/a element)
+                                 ;; {::my-alias/keys [a]}    -> (::my-alias/a element)
+                                 default-ns                     (n/keyword-node (keyword default-ns (name local-sexpr)) default-ns-auto-resolved?)
+                                 ;; {::keys [a]}             -> (::a element)
+                                 default-ns-auto-resolved?      (n/keyword-node (keyword local-sexpr) true)
+                                 ;; {:keys [a]}              -> (:a element)
+                                 :else                          (n/keyword-node (keyword local-sexpr)))))
```

Comment:
thanks for the comments, it helped reviewing this section

Reply:
I couldn't keep track of what was going on 10 minutes after I wrote each of these lines, so I had to add these comments 😂